lib/Dialect/Handshake/HandshakeTypes.cpp

```diff
@@ -99,9 +103,152 @@ checkChannelExtra(function_ref<InFlightDiagnostic()> emitError,
+/// we don't parse square brackets here.
+/// For example: spec: i1, tag: i32
+/// This function is common to both ControlType and ChannelType.
+static LogicalResult
+parseExtraSignals(function_ref<InFlightDiagnostic()> emitError,
+                  AsmParser &odsParser,
+                  SmallVector<ExtraSignal::Storage> &extraSignalsStorage) {
+
+  auto parseSignal = [&]() -> ParseResult {
+    auto &signal = extraSignalsStorage.emplace_back();
+
+    if (odsParser.parseKeywordOrString(&signal.name) ||
+        odsParser.parseColon() || odsParser.parseType(signal.type))
+      return failure();
+
+    // Attempt to parse the optional upstream symbol
+    if (!odsParser.parseOptionalLParen()) {
+      std::string upstreamSymbol;
+      if (odsParser.parseKeywordOrString(&upstreamSymbol) ||
+          upstreamSymbol != UPSTREAM_SYMBOL || odsParser.parseRParen())
+        return failure();
+      // The default value of `signal.downstream` is true.
+      // I think this is a bit confusing.
+      // TODO: signal.upstream?
+      signal.downstream = false;
+    }
+    return success();
+  };
+
+  if (odsParser.parseCommaSeparatedList(parseSignal))
+    return failure();
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// ControlType
+//===----------------------------------------------------------------------===//
+
+void ControlType::print(AsmPrinter &odsPrinter) const {
+  odsPrinter << "<";
+  if (!getExtraSignals().empty()) {
+    printExtraSignals(odsPrinter, getExtraSignals());
+  }
+  odsPrinter << ">";
+}
+
+LogicalResult ControlType::verify(function_ref<InFlightDiagnostic()> emitError,
+                                  ArrayRef<ExtraSignal> extraSignals) {
+  return checkChannelExtra(emitError, extraSignals);
+}
+
+/// Parse the control type after "<[" is parsed.
+static Type parseControlAfterLSquare(AsmParser &odsParser) {
+  function_ref<InFlightDiagnostic()> emitError = [&]() {
+    return odsParser.emitError(odsParser.getCurrentLocation());
+  };
+
+  // Declare vector of structs for storing parse results
+  SmallVector<ExtraSignal::Storage> extraSignalsStorage;
+  if (failed(parseExtraSignals(emitError, odsParser, extraSignalsStorage)))
+    return {};
+
+  // Parse ']'
+  if (odsParser.parseRSquare())
+    return {};
+  // Parse literal '>'
+  if (odsParser.parseGreater())
+    return {};
+
+  SmallVector<ExtraSignal> extraSignals;
+  // Convert parse results to ExtraSignal instances
+  for (const ExtraSignal::Storage &signalStorage : extraSignalsStorage)
+    extraSignals.emplace_back(signalStorage);
+
+  if (failed(checkChannelExtra(emitError, extraSignals)))
+    return {};
+
+  // Build ControlType
+  // Uniquify and Allocate the ExtraSignal instances inside MLIR context
+  return ControlType::get(odsParser.getContext(), extraSignals);
+}
+
+Type ControlType::parse(AsmParser &odsParser) {
+  // Parse literal '<'
```

Comment:
This kind of comments are too fined grained, they don't help readability and clutter the code imo

Reply:
this is your comment, from ChannelType::parse

Reply:
mb, fell free to nuke it then. Sorry if this gave a bad example.